lib/bap_trace/bap_trace.ml

```diff
@@ -0,0 +1,229 @@
+  let fail_on_error   = `Fail
+  let ignore_errors   = `Miss
+  let stop_on_error   = `Stop
+  let warn_on_error f = `Warn f
+  let pack_errors f   = `Pack f
+  let create f        = `User f
+end
+
+module type S = sig
+  val name: string 
+  val supports: 'a tag -> bool
+end
+
+module type P = sig
+  include S
+  val probe: Uri.t -> bool
+end
+
+type io_error = [ 
+  | `Protocol_error of Error.t  (** Data encoding problem         *)
+  | `System_error of Unix.error (** System error                  *)
+] 
+
+type error = [
+  | io_error
+  | `No_provider    (** No provider for a given URI               *)
+  | `Ambiguous_uri  (** More than one provider for a given URI    *)
+] 
+
+module Reader = struct
+  type t = {
+    tool : tool; 
+    meta : dict; 
+    next : unit -> event Or_error.t option;  
+  }
+end
+
+type reader = Reader.t
+
+type t = {
+  id     : id;
+  meta   : dict;
+  events : event seq;
+  tool   : tool;
+  proto  : proto option;
+}
+
+let mk_tab = Tab.create
+let tools  : (module S) Tab.t = mk_tab () 
+let protos : (module P) Tab.t = mk_tab () 
+let readers: (Uri.t -> id -> (reader, io_error) Result.t) Tab.t = mk_tab ()
+let writers: (Uri.t -> t -> (unit, io_error) Result.t) Tab.t = mk_tab ()
+
+let make_id () = Bap_uuid.create `V4
+
+let make_error = function 
+  | #error as r -> Error r
+
+let protocols_of_uri uri = 
+  Hashtbl.fold protos ~init:[] 
+    ~f:(fun ~key ~data acc -> 
+        let module A = (val data : P) in
+        if A.probe uri then key::acc
+        else acc) 
+
+let find_proto uri  = 
+  match protocols_of_uri uri with
+  | [] -> make_error `No_provider
+  | p::[] -> Ok p 
+  | protos -> make_error `Ambiguous_uri 
+
+let find_by_proto tab proto = 
+  try 
+    Ok (Hashtbl.find_exn tab proto)
+  with Not_found -> make_error `No_provider
+
+let make_stream next init monitor = 
+  let open Seq.Step in
+  let if_error m = 
+    fun a -> match next a with 
+      | None -> Done
+      | Some (Ok ev, a') -> Yield (ev, a')
+      | Some (Error err, a') -> 
+        match m with 
+        | `Fail -> Error.raise err
+        | `Miss -> Skip a'
+        | `Pack f -> Yield (f err, a') 
+        | `Stop -> Done
+        | `Warn f -> f err; Skip a' in
+  let ident a = match next a with 
+    | None -> Done
+    | Some (elt, a') -> Yield (elt, a') in
+  let s = match monitor with 
+    | #error_handler as m -> Seq.unfold_step ~init ~f:(if_error m)
+    | `User handle -> Seq.unfold_step ~init ~f:ident |> handle in
+  Seq.memoize s
+
+let wrap_next f = 
+  fun () -> match f () with
+    | None -> None 
+    | Some elt -> Some (elt, ())
+
+let of_reader id proto monitor reader =  
+  let tool = reader.Reader.tool in 
+  let meta = reader.Reader.meta in 
+  let next = wrap_next reader.Reader.next in
+  let events = make_stream next () monitor in
+  {id; meta; tool; events; proto = Some proto;}
+
+let map_result f = function
+  | Ok v -> Ok (f v)
+  | Error err -> make_error err
+  
+let load ?(monitor=`Fail) uri =
```

Comment:
FYI, our usual convention for monadic code is to put lambda's on the same line as binding operator,
```ocaml
find_proto uri >>= fun proto ->
find_proto writers proto >>= fun write ->
write uri t |>
map_result ident
```

In that case each new operation starts at the beginning of a line, thus making the code more readable.